horovod/mxnet/mpi_ops.cc

```diff
@@ -142,50 +149,57 @@ void DoBroadcast(NDArray* tensor, NDArray* output, int root_rank,
     hvd_output = std::make_shared<MXTensor<NDArray>>(output);
   }
 
+  handle_manager.AttachCallback(handle, cb);
+
   auto enqueue_result = EnqueueTensorBroadcast(
       hvd_context, hvd_tensor, hvd_output, root_rank, nullptr,
-      name, device,
-      [cb](const Status& status) {
-        cb();
+      GetOpName("broadcast", name, handle), device,
+      [handle](const Status& status) {
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
       });
   ThrowIfError(enqueue_result);
 }
 
 #if HAVE_CUDA
 void DoBroadcastCudaOnCPU(
     std::shared_ptr<MXTemporaryBuffer<NDArray>>& hvd_cpu_buffer, int root_rank,
-    std::string& name, Callback cb) {
+    const char* name, int handle, Callback cb) {
   // Make async copy of input tensor to CPU tensor and record completion event.
   auto hvd_context = std::make_shared<MXOpContext<NDArray>>(
       CPU_DEVICE_ID, hvd_cpu_buffer->tensor());
   auto ready_event =
       std::make_shared<MXReadyEvent<NDArray>>(hvd_cpu_buffer->tensor());
 
+  handle_manager.AttachCallback(handle, cb);
+
   auto enqueue_result = EnqueueTensorBroadcast(
       hvd_context, hvd_cpu_buffer, hvd_cpu_buffer, root_rank, ready_event,
-      name, CPU_DEVICE_ID,
-      [cb](const Status& status) {
-        cb();
+      GetOpName("broadcast", name, handle), CPU_DEVICE_ID,
+      [handle](const Status& status) {
+        handle_manager.MarkDone(handle, status);
+        hdnale_manager.ExecuteCallback(handle);
```

Comment:
I think we need to call ExecuteCallback first and then MarkDone. Otherwise we may introduce a race here and callback can be skipped.

Reply:
Good catch. This might have caused race condition